File: src/utils/modifier.ts
interface Modifier {
  padding(
    value:
      | number
      | {
          horizontal?: number;
          vertical?: number;
          top?: number;
          bottom?: number;
          start?: number;
          end?: number;
          left?: number;
          right?: number;
        }
  ): Modifier;
  width(value: number): Modifier;
  height(value: number): Modifier;
  size(value: number): Modifier;
  fillMaxHeight(): Modifier;
  fillMaxWidth(): Modifier;
  fillMaxSize(): Modifier;
  border(value: {
    width?: number;
    brush?: string;
    color?: string;
    shape?: string;
  }): Modifier;
  alpha(value: number): Modifier;
  backgroundColor(value: { color: string; shape?: string }): Modifier;
  backgroundBrush(value: {
    brush: string;
    shape?: string;
    alpha?: number;
  }): Modifier;
  clipToBounds(): Modifier;
  shadow(value: {
    elevation?: number;
    shape?: string;
    clip?: boolean;
    ambientColor?: string;
    spotColor?: string;
  }): Modifier;
  zIndex(value: number): Modifier;
}

class ModifierBuilder {
  private modifierArray: Record<string, any>[] = [];

  constructor(existingModifiers?: Record<string, any>[]) {
    if (existingModifiers) {
      this.modifierArray = [...existingModifiers]; // Copy existing modifiers
    }
  }

  padding(
    value:
      | number
      | {
          horizontal?: number;
          vertical?: number;
          top?: number;
          bottom?: number;
          start?: number;
          end?: number;
          left?: number;
          right?: number;
        }
  ): ModifierBuilder {
    if (typeof value === "number") {
      return new ModifierBuilder([...this.modifierArray, { padding: value }]);
    } else if (typeof value === "object" && value !== null) {
      const paddingObject: Record<string, number | undefined> = {};
      if (value.horizontal !== undefined) {
        paddingObject.horizontal = value.horizontal;
      }
      if (value.vertical !== undefined) {
        paddingObject.vertical = value.vertical;
      }
      if (value.top !== undefined) {
        paddingObject.top = value.top;
      }
      if (value.bottom !== undefined) {
        paddingObject.bottom = value.bottom;
      }
      if (value.start !== undefined) {
        paddingObject.start = value.start;
      }
      if (value.end !== undefined) {
        paddingObject.end = value.end;
      }
      if (value.left !== undefined) {
        paddingObject.left = value.left;
      }
      if (value.right !== undefined) {
        paddingObject.right = value.right;
      }

      return new ModifierBuilder([
        ...this.modifierArray,
        { padding: paddingObject },
      ]);
    }

    return this;
  }

  width(value: number): ModifierBuilder {
    return new ModifierBuilder([...this.modifierArray, { width: value }]);
  }

  height(value: number): ModifierBuilder {
    return new ModifierBuilder([...this.modifierArray, { height: value }]);
  }

  size(value: number): ModifierBuilder {
    return new ModifierBuilder([...this.modifierArray, { size: value }]);
  }

  fillMaxHeight(): ModifierBuilder {
    return new ModifierBuilder([
      ...this.modifierArray,
      { fillMaxHeight: true },
    ]);
  }

  fillMaxWidth(): ModifierBuilder {
    return new ModifierBuilder([...this.modifierArray, { fillMaxWidth: true }]);
  }

  fillMaxSize(): ModifierBuilder {
    return new ModifierBuilder([...this.modifierArray, { fillMaxSize: true }]);
  }

  border(value: {
    width?: number;
    brush?: string;
    color?: string;
    shape?: string;
  }): ModifierBuilder {
    return new ModifierBuilder([...this.modifierArray, { border: value }]);
  }

  alpha(value: number): ModifierBuilder {
    return new ModifierBuilder([...this.modifierArray, { alpha: value }]);
  }

  backgroundColor(value: { color: string; shape?: string }): ModifierBuilder {
    return new ModifierBuilder([
      ...this.modifierArray,
      { backgroundColor: value },
    ]);
  }

  backgroundBrush(value: {
    brush: string;
    shape?: string;
    alpha?: number;
  }): ModifierBuilder {
    return new ModifierBuilder([
      ...this.modifierArray,
      { backgroundBrush: value },
    ]);
  }

  clipToBounds(): ModifierBuilder {
    return new ModifierBuilder([...this.modifierArray, { clipToBounds: true }]);
  }

  shadow(value: {
    elevation?: number;
    shape?: string;
    clip?: boolean;
    ambientColor?: string;
    spotColor?: string;
  }): ModifierBuilder {
    return new ModifierBuilder([...this.modifierArray, { shadow: value }]);
  }

  zIndex(value: number): ModifierBuilder {
    return new ModifierBuilder([...this.modifierArray, { zIndex: value }]);
  }

  build(): Record<string, any>[] {
    return this.modifierArray;
  }
}

export const Modifier: Modifier = new ModifierBuilder();

File: android/src/main/java/expo/modules/jetpackcomposereactnative/common/Modifier.kt
package expo.modules.jetpackcomposereactnative.common

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.padding
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clipToBounds
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.RectangleShape
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.zIndex

typealias ModifierProp = List<Map<String, Any>>

fun ModifierProp.toModifier(): Modifier {
    var modifiers: Modifier = Modifier

    forEach { map ->
        map.forEach { (key, value) ->
            when (key) {
                "padding" -> {
                    if (value is Number) {
                        modifiers = modifiers.then(Modifier.padding(value.toDouble().dp))
                    } else if (value is Map<*, *>) {
                        val horizontal = (value["horizontal"] as? Number)?.toDouble()?.dp ?: 0.dp
                        val vertical = (value["vertical"] as? Number)?.toDouble()?.dp ?: 0.dp
                        val top = (value["top"] as? Number)?.toDouble()?.dp ?: vertical
                        val bottom = (value["bottom"] as? Number)?.toDouble()?.dp ?: vertical
                        val start = (value["start"] as? Number)?.toDouble()?.dp ?: horizontal
                        val end = (value["end"] as? Number)?.toDouble()?.dp ?: horizontal

                        modifiers = modifiers.then(
                            Modifier.padding(start, top, end, bottom)
                        )
                    }
                }


                "width" -> {
                    if (value is Number) {
                        modifiers = modifiers.then(Modifier.width(value.toDouble().dp))
                    }
                }

                "height" -> {
                    if (value is Number) {
                        modifiers = modifiers.then(Modifier.height(value.toDouble().dp))
                    }
                }

                "size" -> {
                    if (value is Number) {
                        modifiers = modifiers.then(Modifier.size(value.toDouble().dp))
                    }
                }

                "fillMaxHeight" -> {
                    modifiers = modifiers.then(Modifier.fillMaxHeight())
                }

                "fillMaxWidth" -> {
                    modifiers = modifiers.then(Modifier.fillMaxWidth())
                }

                "fillMaxSize" -> {
                    modifiers = modifiers.then(Modifier.fillMaxSize())
                }

                "border" -> {
                    if (value is Map<*, *>) {
                        val width = value["width"] as? Double ?: 1.0
                        val color = value["color"] as? String ?: "black"

                        modifiers = modifiers.then(
                            Modifier.border(
                                width = width.dp,
                                color = parseComposeColor(color),
                                shape = RectangleShape
                            )
                        )
                    }
                }

                "alpha" -> {
                    if (value is Number) {
                        val float = value.toFloat()
                        modifiers = modifiers.then(Modifier.alpha(float))
                    }
                }

                "backgroundColor" -> {
                    if (value is Map<*, *>) {
                        val color = value["color"] as? String ?: ""
                        modifiers = modifiers.then(
                            Modifier.background(
                                parseComposeColor(color),
                                RectangleShape
                            )
                        )
                    }
                }

                "clipToBounds" -> {
                    modifiers = modifiers.then(Modifier.clipToBounds())
                }

                "shadow" -> {
                    if (value is Map<*, *>) {
                        val elevation = value["elevation"] as? Double ?: 0.0
                        val shape = value["shape"] as? Shape ?: RectangleShape
                        val clip = value["clip"] as? Boolean ?: false
                        val ambientColor = value["ambientColor"] as? String ?: "#000"
                        val spotColor = value["spotColor"] as? String ?: "#000"
                        modifiers = modifiers.then(
                            Modifier.shadow(
                                elevation.dp,
                                shape,
                                clip,
                                parseComposeColor(ambientColor),
                                parseComposeColor(spotColor)
                            )
                        )
                    }
                }

                "zIndex" -> {
                    if (value is Number) {
                        modifiers = modifiers.then(Modifier.zIndex(value.toFloat()))
                    }
                }
            }
        }
    }
    return modifiers
}
